ethereum/contracts/ShutdownSwitch.sol

```diff
@@ -0,0 +1,134 @@
+
+    /// @dev Function that must be implemented by contracts inheriting from this one.
+    function getCurrentGuardianSet() public virtual view returns (Structs.GuardianSet memory);
+
+    /// @dev The number of disable votes required to disable transfers (assuming there are at least that many guardians).
+    uint16 constant REQUIRED_NO_VOTES = 3;
+
+    /// @dev Function that should be called from setup() on new contract deployments, or initialize() on the initial upgrade to deploy this feature.
+    function setUpShutdownSwitch() internal {
+        enabled = true;
+    }
+
+    /// @dev modifier used to block transfers when shutdown.
+    modifier isEnabled {
+        require(enabledFlag(), "transfers are temporarily disabled");
+        _;
+    }
+
+    /// @dev This is the function that allows guardians to vote, and determines the resulting shutdown status.
+    function castShutdownVote(bool _enabled) public {
+        // We always use the current guardian set.
+        Structs.GuardianSet memory gs = getCurrentGuardianSet();
+
+        // Only currently active guardians are allowed to vote.
+        require(isRegisteredVoter(msg.sender, gs), "you are not a registered voter");
+
+        // Only votes to disable are maintained in the map.
+        if (_enabled) {
+            delete disabledVotes[msg.sender];
+        } else {
+            disabledVotes[msg.sender] = true;
+        }
+
+        // Update the number of votes to disable.
+        uint16 votesToDisable = computeNumVotesDisabled(gs);
+
+        // Determine the new shutdown status and generate the appropriate events.
+        bool newEnabledFlag = (votesToDisable < computeRequiredVotesToDisable(gs.keys.length));
+        emit ShutdownSwitch.ShutdownVoteCast(msg.sender, _enabled, votesToDisable, newEnabledFlag);
+
+        if (newEnabledFlag != enabled) {
+            enabled = newEnabledFlag;
+            emit ShutdownSwitch.ShutdownStatusChanged(newEnabledFlag, votesToDisable);
+        }
+    }
+
+    /// @dev Determines if a voter is a current guardian.
+    function isRegisteredVoter(address voter, Structs.GuardianSet memory gs) private pure returns(bool) {
+        for (uint idx = 0; (idx < gs.keys.length); idx++) {
+            if (voter == gs.keys[idx]) {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    /// @dev Returns the number of disable votes required to go disabled.
+    function computeRequiredVotesToDisable(uint numGuardians) private pure returns(uint16) {
+        // If the number of active guardians is less than the pre-configured threshold, use that.
+        return uint16(numGuardians >= REQUIRED_NO_VOTES ? REQUIRED_NO_VOTES : numGuardians);
+    }
+
+    /// @dev Counts up the number of current guardians that have active votes to disable.
+    function computeNumVotesDisabled(Structs.GuardianSet memory gs) private view returns (uint16) {
```

Comment:
I would just call this `getShutdownVoteCount`

Reply:
same, I think it's more direct.

Reply:
Rename it.